### PR TITLE
Track object modifications in fake clientset

### PR DESCRIPTION
**Pull request by olegshaldybin**
Fake clientset is used by unit tests extensively but it has some
shortcomings:
- no filtering on namespace and name: tests that want to test objects in
  multiple namespaces end up getting all objects from this clientset,
  as it doesn't perform any filtering based on name and namespace;
- updates and deletes don't modify the clientset state, so some tests
  can get unexpected results if they modify/delete objects using the
  clientset;
- it's possible to insert multiple objects with the same
  kind/name/namespace, this leads to confusing behavior, as retrieval is
  based on the insertion order, but anchors on the last added object as
  long as no more objects are added.

This change changes core.ObjectRetriever implementation to track object
adds, updates and deletes.

Some unit tests were depending on the previous (and somewhat incorrect)
behavior. These are fixed in the following few commits.

### CODE REVIEW

**Comment by olegshaldybin** [comment]
Any chance someone can take a look, or kick off tests for this branch? We are planning on another PR son, and it depends on the corrected behavior. Would be great to get this in if possible (or get someone's opinion on why it's not a good idea).

**Comment by mikedanese** [comment]
@deads2k @caesarxuchao

File: pkg/client/testing/core/fixture.go [hunk 1]
```diff
@@ -49,153 +58,158 @@ type ObjectScheme interface {
...
 		kind.Version = resource.Version
-		if err != nil {
-			return false, nil, fmt.Errorf("unrecognized action %s: %v", action.GetResource(), err)
-		}
 
-		// TODO: have mapper return a Kind for a subresource?
-		switch castAction := action.(type) {
-		case ListAction:
+		// Here and below we need to switch on implementation types,
+		// no on interfaces, as some interfaces are identical
```
Comment (caesarxuchao) [inline pkg/client/testing/core/fixture.go]: s/no/not

**Comment by olegshaldybin** [comment]
Fixed a typo and rebased.

**Comment by caesarxuchao** [comment]
@olegshaldybin Thanks for the PR. I agree with your first and third points, but I'm not sure about the second point (`updates and deletes don't modify the clientset state`). I think if a test needs to change the objects in the registry, then it should be an integration test that runs a real API server. 
IMO, the fake client should be dumb and should only return pre-programmed objects, and perhaps we should name it `mock` client rather than `fake`. @deads2k what do you think?

**Comment by olegshaldybin** [comment]
@caesarxuchao That makes sense, however there are existing unit tests that rely on modifying objects using the clientset and observing the modified record later on: this can get confusing if they retrieve the old version of the resource after the update. If you want I can try and reduce this change to just handle namespace/name filtering and duplicate insertions, but that can lead to some tests not really testing the real flow (e.g. namespace finalizer test, deployment controller test).

File: pkg/controller/deployment/deployment_controller_test.go [hunk 2]
```diff
@@ -778,18 +773,13 @@ func TestSyncDeploymentCreatesReplicaSet(t *testing.T) {
...
-	updatedRS := newReplicaSet(d, "deploymentrs-4186632231", 1)
-	opt := newListOptions()
+	rs := newReplicaSet(d, "deploymentrs-4186632231", 1)
 
 	f.expectCreateRSAction(rs)
 	f.expectUpdateDeploymentAction(d)
-	f.expectUpdateRSAction(updatedRS)
-	f.expectListPodAction(rs.Namespace, opt)
-	f.expectUpdateDeploymentAction(d)
+	f.expectListPodAction(rs.Namespace, api.ListOptions{})
```
Comment (caesarxuchao) [inline pkg/controller/deployment/deployment_controller_test.go]: This is a nice catch. For a moment I thought the old test was correct, but after some debugging I realized you are right. cc @janetkuo (for she maintains the deployment API).

This is a good example of why a minimum registry in the fake clientset is useful, because *i)*syncDeployment() has multiple interactions with the API server and its behavior is driven by the states stored in the API server, and *ii)* we want to verify the sequence of the client requests, which currently cannot be done in an integration test.

An alternative solution is running this as an integration test, and add a wrapper around a real clientset to record all the requests sent. cc @lavalamp.

**Comment by janetkuo** [comment]
ok to test

**Comment by deads2k** [comment]
> I think if a test needs to change the objects in the registry, then it should be an integration test that runs a real API server.

I find @olegshaldybin's argument pretty compelling.  I'll review the pull today, but if the code looks reasonably simple I think I'd be in favor of allowing it.

File: pkg/client/testing/core/fixture.go [hunk 3]
```diff
@@ -49,153 +58,158 @@ type ObjectScheme interface {
...
+// ObjectReaction returns a ReactionFunc that applies core.Action to
+// the given registry.
+func ObjectReaction(registry Registry, mapper meta.RESTMapper) ReactionFunc {
 	return func(action Action) (bool, runtime.Object, error) {
+		ns := action.GetNamespace()
 		resource := action.GetResource()
+
 		kind, err := mapper.KindFor(resource)
+		if err != nil {
+			return false, nil, fmt.Errorf("unrecognized action for resource %s: %s", resource, err)
```
Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: "no kind for registered for resource"?  I know its pre-existing, but you've already tweaked it some.

File: pkg/client/testing/core/fixture.go [hunk 4]
```diff
@@ -18,28 +18,37 @@ package core
...
-	// Add adds a runtime object for test purposes into this object.
+// Registry keeps track of objects. It is intended to be used to fake
+// calls to a server by returning objects based on their kind,
+// namespace and name.
+type Registry interface {
+	// Get retrieves the object by its kind, namespace and name. If
+	// kind is a list, all objects that match ns and name are returned
+	// in their corresponding list enclosure (e.g. PodList,
+	// ServiceList).
+	Get(gvk unversioned.GroupVersionKind, ns, name string) (runtime.Object, error)
```
Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: Can we get a separate list method?

File: pkg/client/testing/core/fixture.go [hunk 5]
```diff
@@ -49,153 +58,158 @@ type ObjectScheme interface {
...
-			return true, resource, err
+		case CreateActionImpl:
+			objMeta, err := meta.Accessor(typedAction.GetObject())
+			if err != nil {
+				return true, nil, err
+			}
+			if typedAction.GetSubresource() == "" {
+				err = registry.Add(typedAction.GetObject())
+			} else {
+				// Handle subresource creation as an update on the
```
Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: Can you model this as a separate method that takes the subresource and the object and must return the correct value to you?  Some subresource posts require special handling.  I can certainly see making a standard implementation for handling "status" since its so common.

File: pkg/client/testing/core/fixture.go [hunk 6]
```diff
@@ -49,153 +58,158 @@ type ObjectScheme interface {
...
-	return nil
-}
-
-type objects struct {
-	types   map[string][]runtime.Object
-	last    map[string]int
+type registry struct {
 	scheme  ObjectScheme
 	decoder runtime.Decoder
+	mu      sync.RWMutex
```
Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: `lock`?

File: pkg/client/testing/core/fixture.go [hunk 7]
```diff
@@ -49,153 +58,158 @@ type ObjectScheme interface {
...
+	objects map[string][]runtime.Object
```
Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: doc describing key/value.  I was expecting `map[string]map[string]runtime.Object` for namespace, name, object.

File: pkg/client/testing/core/fixture.go [hunk 8]
```diff
@@ -49,153 +58,158 @@ type ObjectScheme interface {
...
-	nilValue := reflect.Zero(reflect.TypeOf(empty)).Interface().(runtime.Object)
 
-	arr, ok := o.types[kind.Kind]
+	out, err := r.scheme.New(kind)
+	if err != nil {
+		return nil, err
+	}
+
+	nilValue := reflect.Zero(reflect.TypeOf(out)).Interface().(runtime.Object)
+	isList := strings.HasSuffix(kind.Kind, "List")
```
Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: You should be able to do `meta.IsListType(scheme.New(kind))`

File: pkg/client/testing/core/fixture.go (hunk 7 above)
Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: Actually, I think this should be 
`map[unversioned.GroupResource]map[string(namespace)]map[string(name)]runtime.Object`  I don't think you should index by kind, since we access by resource.

File: pkg/client/testing/core/fixture.go [hunk 9]
```diff
@@ -209,41 +223,124 @@ func (o objects) Kind(kind unversioned.GroupVersionKind, name string) (runtime.O
 	return out, nil
 }
 
-func (o objects) Add(obj runtime.Object) error {
-	gvks, _, err := o.scheme.ObjectKinds(obj)
+func (r *registry) Add(obj runtime.Object) error {
+	return r.add(obj, false)
```
Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: Check for existence before starting, you should produce an "already exists" error.

File: pkg/client/testing/core/fixture.go [hunk 10]
```diff
@@ -209,41 +223,124 @@ func (o objects) Kind(kind unversioned.GroupVersionKind, name string) (runtime.O
...
+}
+
+func (r *registry) Update(obj runtime.Object) error {
+	return r.add(obj, true)
```
Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: check for existence before starting, you should produce a "not found" error.

File: pkg/client/testing/core/fixture.go [hunk 11]
```diff
@@ -209,41 +223,124 @@ func (o objects) Kind(kind unversioned.GroupVersionKind, name string) (runtime.O
...
+
+func (r *registry) add(obj runtime.Object, replaceExisting bool) error {
+	gvks, _, err := r.scheme.ObjectKinds(obj)
 	if err != nil {
 		return err
 	}
+	if len(gvks) == 0 {
+		return fmt.Errorf("no registered object kind for %v", obj)
+	}
+	groupResource := unversioned.GroupResource{Group: gvks[0].Group, Resource: gvks[0].Kind}
```
Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: You'll need to plumb the actual resource down.

**Comment by deads2k** [comment]
I like it in concept, but the implementation needs to keep the distinction between kind and resource clear.  Looking at what's here, I think the "update" and "add" affecting "get" and "list" looks like a pretty natural outgrowth.

**Comment by olegshaldybin** [comment]
@deads2k Thanks for you review! I was trying to minimize the actual code change for the additions but most of your feedback makes sense, I'll go ahead and incorporate it.

File: pkg/client/testing/core/fixture.go [hunk 12]
```diff
@@ -49,153 +58,158 @@ type ObjectScheme interface {
...
-// subsequent calls.
-func NewObjects(scheme ObjectScheme, decoder runtime.Decoder) ObjectRetriever {
-	return objects{
-		types:   make(map[string][]runtime.Object),
-		last:    make(map[string]int),
+var _ Registry = &registry{}
+
+// NewObjects returns a Registry that can be used to keep track of
+// objects for the fake clientset. Mostly useful for unit tests.
+func NewObjects(scheme ObjectScheme, decoder runtime.Decoder) Registry {
```
Comment (caesarxuchao) [inline pkg/client/testing/core/fixture.go]: Rename to NewRegistry?

File: pkg/client/testing/core/fixture.go (hunk 9 above)
Comment (caesarxuchao) [inline pkg/client/testing/core/fixture.go]: I think the check is done in `add()`.

Comment (olegshaldybin) [inline pkg/client/testing/core/fixture.go]: add() already does that: if existing object matches kind, namespace and name of the one being inserted it returns an "already exists" error.

File: pkg/client/testing/core/fixture.go (hunk 10 above)
Comment (olegshaldybin) [inline pkg/client/testing/core/fixture.go]: add() does that as well, if replaceExisting flag is true but no existing object found it returns a "not found" error.

File: pkg/client/testing/core/fixture.go (hunk 7 above)
Comment (olegshaldybin) [inline pkg/client/testing/core/fixture.go]: Should it be `GroupResource` or `GroupVersionKind`? It seems natural to bucket them by `GroupVersionKind`, since that's what we can extract from the object being added. I'd like to understand more about the difference between `Resource` and `Kind`.

File: pkg/client/testing/core/fixture.go (hunk 12 above)
Comment (olegshaldybin) [inline pkg/client/testing/core/fixture.go]: I would love to, but that will require regenerating a fake clientset code. Do you think it's worth it?

File: pkg/client/testing/core/fixture.go (hunk 4 above)
Comment (olegshaldybin) [inline pkg/client/testing/core/fixture.go]: Yeah, it will make things cleaner for sure. I am still not clear on a distinction between GroupResource and GroupKind: existing code seems to be using them interchangeably.

**Comment by janetkuo** [comment]
@k8s-bot ok to test

**Comment by caesarxuchao** [comment]
Could you add a comment to `NewSimpleClientset` to say something like "the returned fake clientset by default uses a fake registry, which may not reflect the real behavior of the API server (e.g., no validation or defaulting), and if the user wish to *not* use the fake registry, they can specify a different Reactor"?

**Comment by olegshaldybin** [comment]
@caesarxuchao Sure, will do.

**Comment by caesarxuchao** [comment]
`NewSimpleClientset` is [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/clientset_generated/release_1_3/fake/clientset_generated.go#L39). That will also require regenerating the clientset.

The template is [here](https://github.com/kubernetes/kubernetes/blob/master/cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_clientset.go#L115), you just need to change the template and then run hack/update-codegen.sh.

**Comment by lavalamp** [comment]
I would kind of prefer it if you did not get a fake registry by default, if
you had to invoke something special to get it. This is because I don't
expect it to perfectly replicate a real registry (validation, etc) and I
don't want tests to unintentionally assume that it does.

**Comment by olegshaldybin** [comment]
@lavalamp But it seems to me even worse if tests assume that their previous modification actually went through because client returned no error, but getting the old version of the object on the next request, since the registry only contains objects explicitly added by the test framework.

**Comment by lavalamp** [comment]
It depends on what sort of a test you're writing. If you want to verify an
exact sequence of calls (a form of test that IMO is typically bad outside
of special circumstances) then you should expect that behavior, it's how
all the mocking libraries I'm aware of work.

If you want to verify some other behavior, like a correct result after a
sequence of calls, then you probably want a real integration test, with a
real registry. Otherwise what are you really verifying?

**Comment by olegshaldybin** [comment]
@lavalamp I understand your concern, but in that case only read operations should be allowed against this default reactor, and any modifications should return an error. Otherwise it's easy to write unit tests that don't reflect the actual system behavior (you can see some examples in this PR).

I can reduce the change to just apply namespace filtering properly if that's the consensus.

File: pkg/controller/deployment/deployment_controller_test.go (hunk 2 above)
Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller_test.go]: Thanks, this is correct. 
The controller first syncs the deployment and then creates an rs with 1 replica immediately (`expectCreateRSAction`), update the deployment's revision (`expectUpdateDeploymentAction`), and then list pods to calculate deployment status (`expectListPodAction`).

**Comment by caesarxuchao** [comment]
> I would kind of prefer it if you did not get a fake registry by default, if you had to invoke something special to get it.

@olegshaldybin could you do this? Maybe adding a `bool fakeRegisty` flag to `NewSimpleClientset` and fall back to the old behavior if the flag is false (this could be messy..)? This will make sure people understand they are not interacting with a real registry.
@lavalamp do you think adding a comment stating "the fake registry is not an accurate replica of the real registry" to the `fake.NewSimpleClientset` is enough?

**Comment by olegshaldybin** [comment]
@caesarxuchao Flag is doable but it can get pretty messy. I'll try and experiment with it though.

File: pkg/client/testing/core/fixture.go (hunk 7 above)
Comment (caesarxuchao) [inline pkg/client/testing/core/fixture.go]: I would say it should be `GroupVersionResource` here? Why `GroupResource`? 

> It seems natural to bucket them by GroupVersionKind, since that's what we can extract from the object being added

Actually I think we shouldn't extract it from the object. We should use the `GroupVersionResource` populated from the `Action`.

> I'd like to understand more about the difference between Resource and Kind

`GroupVersionResource` indicates what URL the request will be sent to. `GroupVersionKind` is a property of the object. In most cases they are the same, but sometimes GVK can be translated to different GVRs (e.g. `Scale`). Also, GVK.Version can be empty or "__internalVersion", because we have an internal version for objects, but GVR should never have an empty version, because there is always a version in the URL in k8s.

Comment (olegshaldybin) [inline pkg/client/testing/core/fixture.go]: Thanks, that's very helpful.

Comment (olegshaldybin) [inline pkg/client/testing/core/fixture.go]: @caesarxuchao Here's one thing I am not sure how to implement: say we want to have a `List()` method that receives a `GroupVersionResource` argument: now we want to create a list of objects of a certain kind. How do I translate the incoming `GroupVersionResource` to a `GroupVersionKind` of the correct `List` object?

Comment (olegshaldybin) [inline pkg/client/testing/core/fixture.go]: Ah, I think I can do it by using `mapper.KindFor`.

Comment (caesarxuchao) [inline pkg/client/testing/core/fixture.go]: @olegshaldybin could you point me to the code where you need to create the list of objects? I must miss something, because I think we don't need to get `GVK`. I hope we don't need to use `mapper` in the fake clientset.

Comment (olegshaldybin) [inline pkg/client/testing/core/fixture.go]: `GVK` is actually used all over the place in the existing implementation. Here's where it creates the  result object based on a `GVK` passed in:

https://github.com/kubernetes/kubernetes/blob/master/pkg/client/testing/core/fixture.go#L17

It's already using mapper to retrieve `GVK` from a `GVR`:

https://github.com/kubernetes/kubernetes/blob/master/pkg/client/testing/core/fixture.go#L61

I am definitely interested in improving some of these patterns, but not sure if we can avoid using `GVK`. To be fair, I don't have a full understanding of k8s API machinery yet, just started looking at how it's organized, so somewhat hesitant to change it too much. But if you have some pointers to a better approach, I can take a stab at it.

Comment (olegshaldybin) [inline pkg/client/testing/core/fixture.go]: Also, when setting tests up, objects are being added to the clientset directly, at that point we don't know what `GroupVersionResource` to associate them with, since we don't have any REST context at that point, just a `runtime.Object`.

Comment (caesarxuchao) [inline pkg/client/testing/core/fixture.go]: I haven't looked carefully, but I think we can get rid of GVK in most places, but we still need GVK where we called [o.scheme.New()](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/testing/core/fixture.go#L161) to get List objects and empty objects.

I'm keen on getting rid of calls to mapper.Kind() because that won't work once we convert controllers to use versioned clientsets, because mapper.Kind() always return the internal kind, while the controllers will expect to receive versioned Lists. Anyway, this is a separate issue, use GVK where you have to and we'll deal with it later.

Comment (caesarxuchao) [inline pkg/client/testing/core/fixture.go]: > Also, when setting tests up, objects are being added to the clientset directly, at that point we don't know what GroupVersionResource to associate them with, since we don't have any REST context at that point, just a runtime.Object.

Right. Perhaps we can ask the caller to provide GVR along with the objects, but that will be a separate PR.

**Comment by olegshaldybin** [comment]
@caesarxuchao @deads2k Sorry, I just got back to this change today. Responded to most feedback and renamed Registry to ObjectTracker to avoid any confusion with real registries. Also clarified the intent in NewSimpleClientset.

Changing internal tracking to be based `GroupVersionResource` probably needs to be its own PR, as it's not straightforward and requires changes to all tests that use the clientset, as currently they don't supply the `GroupVersionResource`, just objects themselves. With the way it's currently used by tests it makes more sense to bucket them by `GVK`. Note that `GVR/GVK` distinction isn't handled by current code, so this change doesn't really make it any worse.

I do share the sentiment that this clientset is not good enough in the long term: we need a better way to mock/record client calls in tests. The pre-existing behavior however is very confusing, so may I suggest taking this change to fix the existing tests (and add filtering capabilities for any future tests that may need it), and plan a better overall approach for the fake clientset in the future.

File: cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_clientset.go [hunk 13]
```diff
@@ -111,9 +111,12 @@ func (g *genClientset) GenerateType(c *generator.Context, t *types.Type, w io.Wr
 
 // This part of code is version-independent, unchanging.
 var common = `
-// Clientset returns a clientset that will respond with the provided objects
+// NewSimpleClientset returns a clientset that will respond with the provided objects.
+// It's backed by a very simple object tracker that processes creates and updates as-is,
```
Comment (caesarxuchao) [inline cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_clientset.go]: Shouldn't it be "creates, updates and deletions"? :)

Comment (olegshaldybin) [inline cmd/libs/go2idl/client-gen/generators/fake/generator_fake_for_clientset.go]: Yes, good point.

**Comment by caesarxuchao** [comment]
@olegshaldybin I'm ok with using `GVK`, and I'm ok with the clarification in the comment of `NewSimpleClientset`. I haven't reviewed the code carefully though, I'll leave that to @deads2k. Thank you both.

**Comment by olegshaldybin** [comment]
@caesarxuchao Thanks! @deads2k, note that I dind't change linear search on an `objects` map to using a map of maps of maps, as it was slightly more code and given the number of objects inn these clientsets (usually <10), O(n) should do fine. If you want I can change to use maps, but it's a little less readable that way.

File: pkg/client/testing/core/fixture.go [hunk 14]
```diff
@@ -49,201 +59,310 @@ type ObjectScheme interface {
...
+			t.objects[gvk] = append(t.objects[gvk][:i], t.objects[gvk][i+1:]...)
+			return nil
 		}
-		o.types[kind] = append(o.types[kind], obj)
 	}
 
 	return nil
 }
 
+func filterByNamespaceAndName(objs []runtime.Object, ns, name string) ([]runtime.Object, error) {
```
Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: doc the meaning of empty here.  Empty namespace is one of the worst constructs we have.  Sometimes it is none, sometimes it is all, sometimes it is any, and sometimes it is "default".

File: pkg/client/testing/core/fixture.go [hunk 15]
```diff
@@ -49,201 +59,310 @@ type ObjectScheme interface {
...
-	index := o.last[kind.Kind]
-	if index >= len(arr) {
-		index = len(arr) - 1
+	matchingObjs, err := filterByNamespaceAndName(objs, ns, "")
+	if err != nil {
+		return nil, err
 	}
-	if index < 0 {
-		return nilValue, errors.NewNotFound(unversioned.GroupResource{Group: kind.Group, Resource: kind.Kind}, name)
+	if len(matchingObjs) == 0 {
```
Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: `SetList` ought to handle empty list.  Does it not?

File: pkg/client/testing/core/fixture.go [hunk 16]
```diff
@@ -49,201 +59,310 @@ type ObjectScheme interface {
...
+	objs, ok := t.objects[gvk]
+	if !ok {
+		return nil, errNotFound
+	}
+
+	matchingObjs, err := filterByNamespaceAndName(objs, ns, name)
+	if err != nil {
+		return nil, err
+	}
+	if len(matchingObjs) == 0 {
```
Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: make an error of some kind for getting more than one here.

File: pkg/client/testing/core/fixture.go [hunk 17]
```diff
@@ -49,201 +59,310 @@ type ObjectScheme interface {
...
+		return nil, err
+	}
+	if list, err = t.scheme.Copy(list); err != nil {
+		return nil, err
+	}
+	return list, nil
+}
+
+func (t *tracker) Get(gvk unversioned.GroupVersionKind, ns, name string) (runtime.Object, error) {
+	obj, err := t.scheme.New(gvk)
```
Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: Use the scheme to figure out if you need to require a namespace or not.  This should not allow a get of a namespaced resource without a namespace.

Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: I'm not seeing why you need this here.  Doesn't `Copy` already return it to you?

File: pkg/client/testing/core/fixture.go [hunk 18]
```diff
@@ -49,201 +59,310 @@ type ObjectScheme interface {
...
 }
 
-func (o objects) Add(obj runtime.Object) error {
-	gvks, _, err := o.scheme.ObjectKinds(obj)
+func (t *tracker) Update(obj runtime.Object) error {
+	return t.add(obj, true)
+}
+
+func (t *tracker) add(obj runtime.Object, replaceExisting bool) error {
+	if meta.IsListType(obj) {
```
Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: How and why would I get here with a list type?

File: pkg/client/testing/core/fixture.go [hunk 19]
```diff
@@ -49,201 +59,310 @@ type ObjectScheme interface {
...
+
+	gvks, _, err := t.scheme.ObjectKinds(obj)
 	if err != nil {
 		return err
 	}
-	kind := gvks[0].Kind
+	if len(gvks) == 0 {
+		return fmt.Errorf("no registered kinds for %v", obj)
+	}
+	gvk := gvks[0]
```
Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: Seems like you should loop through and add it to all of them.  Hitting an error would fail the call and leave this instance in an unknown state, that's ok.

File: pkg/client/testing/core/fixture.go (hunk 18 above)
Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: > How and why would I get here with a list type?

Oh, you're using this for initialization?  I think I'd suggest a different external method for initialization that breaks it apart and does the individual adds.

File: pkg/client/testing/core/fixture.go [hunk 20]
```diff
@@ -49,201 +59,310 @@ type ObjectScheme interface {
...
-		if status, ok := obj.(*unversioned.Status); ok && status.Details != nil {
-			kind = status.Details.Kind
+		if objMeta.GetNamespace() == ns && objMeta.GetName() == name {
+			t.objects[gvk] = append(t.objects[gvk][:i], t.objects[gvk][i+1:]...)
+			return nil
 		}
-		o.types[kind] = append(o.types[kind], obj)
 	}
 
 	return nil
```
Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: I should get a NotFound if I can't delete.

File: pkg/client/testing/core/fixture.go (hunk 15 above)
Comment (olegshaldybin) [inline pkg/client/testing/core/fixture.go]: It does, I'll remove the length-check.

File: pkg/client/testing/core/fixture.go (hunk 17 above)
Comment (olegshaldybin) [inline pkg/client/testing/core/fixture.go]: You're right, that's some kind of leftover, will remove.

Comment (olegshaldybin) [inline pkg/client/testing/core/fixture.go]: Not sure how to use scheme to determine whether object is namespaced. `ObjectScheme` implements `ObjectCreater`, `ObjectCopier` and `ObjectTyper`, none of them seem to have any metadata about the object that we can use to check whether it's namespaced.

File: pkg/client/testing/core/fixture.go (hunk 18 above)
Comment (olegshaldybin) [inline pkg/client/testing/core/fixture.go]: Yeah, the existing code assumes Add can accept the list, I wasn't a big fan but decided to keep it. Will experiment with adding a separate `Init()` method.

File: pkg/client/testing/core/fixture.go (hunk 19 above)
Comment (olegshaldybin) [inline pkg/client/testing/core/fixture.go]: Makes sense, will do that.

File: pkg/client/testing/core/fixture.go (hunk 20 above)
Comment (olegshaldybin) [inline pkg/client/testing/core/fixture.go]: Sounds good, will fix.

File: pkg/client/testing/core/fixture.go (hunk 18 above)
Comment (olegshaldybin) [inline pkg/client/testing/core/fixture.go]: It seems that will be slightly more involved than I expected, since tests use both list and non-list objects for the fake clientset initialization and sometimes mix them together. Should I just change all tests to not add lists? The alternative is to keep the current behavior for now and revisit it along with further fake clientset improvements.

Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: > It seems that will be slightly more involved than I expected, since tests use both list and non-list objects for the fake clientset initialization and sometimes mix them together. Should I just change all tests to not add lists? The alternative is to keep the current behavior for now and revisit it along with further fake clientset improvements.

I won't block the pull on this.  Fixing in the future would be desireable.  Initialization is special and I'd rather have accidents in unit tests.

File: pkg/client/testing/core/fixture.go (hunk 17 above)
Comment (deads2k) [inline pkg/client/testing/core/fixture.go]: > Not sure how to use scheme to determine whether object is namespaced. ObjectScheme implements ObjectCreater, ObjectCopier and ObjectTyper, none of them seem to have any metadata about the object that we can use to check whether it's namespaced.

Oh, it moved.  `pkg/apimachiner/registered.Group` will give you `GroupMeta` which has a `RESTMapper`, which can give you the `RESTMapping`, which has the `RESTScope`.

Comment (olegshaldybin) [inline pkg/client/testing/core/fixture.go]: I added the check and it caught even more tests that don't set the namespace for namespaced resources. Will fix those and update the PR once they all pass.

**Comment by olegshaldybin** [comment]
@deads2k I've responded to your feedback, please see last 4 commits. Had to fix some tests after adding the namespace check. Also, noticed that Event objects are namespaced but controllers don't fill in namespace field when they send events, so added a bypass for Events in the namespace check. PTAL.

**Comment by olegshaldybin** [comment]
@deads2k Just rebased so it doesn't get too stale.

**Comment by deads2k** [comment]
lgtm

Can you squash it down to something around four commits?  Something like
 1. new reaction functions and tests for the harness
 1. generated code
 1. intent bugs in tests (looked like deployments may have been bugged?)
 1. "boring" test updates

To make it a little easier to find and track problems.

**Comment by olegshaldybin** [comment]
@deads2k Squashed to 7 commits: object tracker modifications, generated code, commit per functional test fix + one commit for "boring" test fixes. Can probably squash functional test fixes together but it seemed to be slightly cleaner to have them fixed separately.

**Comment by olegshaldybin** [comment]
@deads2k Had to rebase.

**Comment by deads2k** [comment]
The queue isn't open to non-1.3 work right now.  Mention me again after its open to remind me to rescan and tag?

**Comment by olegshaldybin** [comment]
Sounds good.

**Comment by deads2k** [comment]
> Sounds good.

The queue is open again, but this needs a rebase.

**Comment by olegshaldybin** [comment]
@deads2k Thanks, just rebased and fixed merge conflicts. Note that I had to further fix the deployment controller test, since controller logic has changed, and not it also updates deployment status as a part of replica set creation flow (cc @janetkuo).
